crates/snapbox/src/data.rs

```diff
@@ -242,7 +315,62 @@ impl Normalize for NormalizeMatches<'_> {
                     .normalize(&text, &self.pattern.render().unwrap());
                 Data::text(lines)
             }
+            #[cfg(feature = "json")]
+            DataInner::Json(value) => {
+                let mut value = value;
+                if let DataInner::Json(exp) = &self.pattern.inner {
+                    normalize_value_matches(&mut value, exp, self.substitutions);
+                }
+                Data::json(value)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "structured-data")]
+fn normalize_value(value: &mut serde_json::Value, op: fn(&str) -> String) {
+    match value {
+        serde_json::Value::String(str) => {
+            *str = op(str);
         }
+        serde_json::Value::Array(arr) => {
+            arr.iter_mut().for_each(|value| normalize_value(value, op));
+        }
+        serde_json::Value::Object(obj) => {
+            obj.iter_mut()
+                .for_each(|(_, value)| normalize_value(value, op));
+        }
+        _ => {}
+    }
+}
+
+#[cfg(feature = "structured-data")]
+fn normalize_value_matches(
+    actual: &mut serde_json::Value,
+    expected: &serde_json::Value,
+    substitutions: &crate::Substitutions,
+) {
+    use serde_json::Value::*;
+    match (actual, expected) {
+        // "{...}" is a wildcard
+        (act, String(exp)) if exp == "{...}" => {
+            *act = serde_json::json!("{...}");
+        }
+        (String(act), String(exp)) => {
+            *act = substitutions.normalize(act, exp);
+        }
+        (Array(act), Array(exp)) => {
+            act.iter_mut()
+                .zip(exp)
+                .for_each(|(a, e)| normalize_value_matches(a, e, substitutions));
+        }
+        (Object(act), Object(exp)) => {
+            act.iter_mut()
+                .zip(exp)
+                .filter(|(a, e)| a.0 == e.0)
+                .for_each(|(a, e)| normalize_value_matches(a.1, e.1, substitutions));
+        }
+        (_, _) => {}
```

Comment:
I think this is good enough for merge right now.

In the future, we will want to consider
- Should we have a wildcard key for matching all key-value pairs that are not explicitly in the expected
- We will want `"{...}"` to match multiple items in an array as I expect that will be important for ndjson support